netrd/distance/graph_diffusion_distance.py

```diff
@@ -0,0 +1,156 @@
+        `laplacian matrices`, and the output of the optimization process
+        (`peak_diffusion_time` and `peak_deviation`).
+
+        Adapted from the authors' MATLAB code, available at: https://rb.gy/txbfrh
+
+
+        Parameters
+        ----------
+
+        G1, G2 (nx.Graph)
+            two networkx graphs to be compared.
+
+        thresh (float)
+            minimum value above which the eigenvalues will be considered.
+
+        resolution (int)
+            number of :math:`t` values to span through.
+
+        Returns
+        -------
+        dist (float)
+            the distance between `G1` and `G2`.
+
+        References
+        ----------
+
+        .. [1] Hammond, D. K., Gur, Y., & Johnson, C. R. (2013, December).
+               Graph diffusion distance: A difference measure for weighted graphs based on the
+               graph Laplacian exponential kernel. In Global Conference on Signal and
+               Information Processing, 2013 IEEE (pp 419-422). IEEE.
+               https://doi.org/10.1109/GlobalSIP.2013.6736904
+
+        """
+
+        A1 = nx.to_numpy_array(G1)
+        A2 = nx.to_numpy_array(G2)
+
+        L1 = laplacian(A1)
+        L2 = laplacian(A2)
+
+        D1, V1 = np.linalg.eig(L1)
+        D2, V2 = np.linalg.eig(L2)
+
+        eigs = np.hstack((np.diag(D1), np.diag(D2)))
+        eigs = eigs[np.where(eigs > thresh)]
+        eigs = np.sort(eigs)
+
+        if len(eigs) == 0:
+            dist = 0
+            self.results['dist'] = dist
+            return dist
+
+        t_upperbound = np.real(1.0 / eigs[0])
+        ts = np.linspace(0, t_upperbound, resolution)
+
+        # Find the Frobenius norms between all the diffusion kernels at
+        # different times. Return the value and where this vector is minimized.
+        E = -gdd_xi_t(V1, D1, V2, D2, ts)
+        f_val, t_star = (np.nanmin(E), np.argmin(E))
+
+        dist = np.sqrt(-f_val)
+
+        self.results['adjacency_matrices'] = A1, A2
+        self.results['laplacian_matrices'] = L1, L2
+        self.results['peak_diffusion_time'] = t_star
+        self.results['peak_deviation'] = f_val
+
+        self.results['dist'] = dist
+
+        return dist
+
+
+def gdd_xi_t(V1, D1, V2, D2, ts):
+    """
+    Computes Frobenius norm of difference of Laplacian exponential diffusion
+    kernels, at specified timepoints.
+
+    Parameters
+    ----------
+
+    V1, V2 (np.array)
+        eigenvectors of the Laplacians of `G1` and `G2`
+
+    D1, D2 (np.array)
+        eigenvalues of the Laplacians of `G1` and `G2`
+
+    t (float)
+        time at which to compute the difference in Frobenius norms
+
+    Returns
+    -------
+
+    E (np.array)
+        same size as :math:`t`, contains differences of Frobenius norms
+
+    """
+
+    E = np.zeros(len(ts))
+
+    for kt, t in enumerate(ts):
+        ed1 = np.diag(np.exp(-t * np.diag(D1)))
+        ed2 = np.diag(np.exp(-t * np.diag(D2)))
+
+        tmp1 = V1.dot(np.atleast_2d(ed1).T * V1.T)
```

Comment:
Question: Does `ed1.T * V1.T` compute the dot product between two matrices? Vectors?

Reply:
Haven't resolved this. @jkbren can you clarify this since you have more familiarity with the method?

Reply:
`ed1` is a vector and `V1` is matrix. `np.atleast_2d(ed1).T` is turning it into a column vector.

Reply:
OK. I still don't like the usage of `*` for matrix-vector multiplication. Is this recommended? I've never used it myself.

Reply:
I think I see how to clean it up, will investigate more later today.

Reply:
```python
>>> import numpy as np
>>> a = np.array([1, 2, 3, 4])
>>> b = np.arange(16).rehsape((4,4))
>>> a*b
array([[ 0,  2,  6, 12],
       [ 4, 10, 18, 28],
       [ 8, 18, 30, 44],
       [12, 26, 42, 60]])
>>> np.atleast_2d(a).T * b.T 
array([[ 0,  4,  8, 12],
       [ 2, 10, 18, 26],
       [ 6, 18, 30, 42],
       [12, 28, 44, 60]])
>>> (a * b).T
array([[ 0,  4,  8, 12],
       [ 2, 10, 18, 26],
       [ 6, 18, 30, 42],
       [12, 28, 44, 60]])
```

It's element-wise multiplication along the outer axis. The `(a * b).T` notation is a bit cleaner but still doesn't make the vector-matrix multiplication all that explicit. We could use [np.multiply](https://numpy.org/doc/stable/reference/generated/numpy.multiply.html) to make it more explicit, e.g. `np.multiply(a, b).T`; that way it doesn't look like some other form of matrix multiplication, but it still leaves what it's doing a little mysterious.

Reply:
> We could use np.multiply to make it more explicit, e.g. np.multiply(a, b).T

Let's do this.